app/routes.py

```diff
@@ -1,2 +1,304 @@
+def handle_one_task(task_id):
+    task = Task.query.get(task_id)
+    # if taskdoesnt exist, give 404
+    if task is None:
+        return make_response(f"Task {task_id} not found", 404)
+
+    # acconting for the descrepancy in wave 6 where it was saying to implement goal_id in thisresponse body.
+    # fixed this by essentially "if there is a goal id at this task, then return that, if not, then return it without the goal id
+    elif request.method == "GET":
+        if not task.goal_id:
+            return {"task": {
+                "id": task.task_id,
+                "title": task.title,
+                "description": task.description,
+                "is_complete" : (False
+                if task.completed_at == None else True)
+                }},200
+        else:
+            return {"task": {
+            "id": task.task_id,
+            "goal_id": task.goal_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete" : (False
+            if task.completed_at == None else True)
+            }},200
+
+    elif request.method == "PUT":
+        request_body = request.get_json()
+        # returns error message ifnot all fields filled in
+        if "title" not in request_body or "description" not in request_body:
+            return { "message: Requires a title and description and completion status"}, 400
+        task.title = request_body["title"]
+        task.description = request_body["description"]
+        
+
+        # save action and return what you created
+        db.session.add(task)
+        db.session.commit()
+        return {"task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete" : (False
+            if task.completed_at == None else True)
+            }},200
+    # delete task
+    elif request.method == "DELETE":
+        db.session.delete(task)
+        db.session.commit()
+        return {"details":f'Task {task.task_id} "{task.title}" successfully deleted'}
+
+
+@tasks_bp.route("/<task_id>/mark_complete", methods = ["PATCH"])
+def complete_task(task_id):
+    task = Task.query.get(task_id)
+    
+    # return error message if trying to complete task that doesnt exist
+    if task is None:
+            return make_response("", 404)
+    
+    
+    if request.method == "PATCH":
+        
+        # wave 4 slack bot, since we are using GET and it has to do with completing task, put here
+        task.completed_at = datetime.now()
+        slack_url = 'https://slack.com/api/chat.postMessage'
+        slack_params = {
+            "channel": "slack-api-test-channel",
+            "text": f"Someone just completed the task {task.title}"
+        }
+        slack_header = {
+            "Authorization": f"Bearer {os.environ.get('SLACK_API_TOKEN')}"
+        }
+        requests.get(url = slack_url, params = slack_params, headers = slack_header)
```

Comment:
To practice the single responsibility rule, we can move the slack bot logic into a helper function.